File: templates/backend/nodejs/levels/level_2/files/backend/src/routes/todos.js
const { Router } = require("express");

const router = Router();
const todos = new Map();
let nextId = 1;

router.get("/todos", (_req, res) => res.json(Array.from(todos.values())));

router.post("/todos", (req, res) => {
  if (typeof req.body.title !== "string" || req.body.title.trim() === "" || req.body.title.length > 120) {
    res.status(400).json({ detail: "Todo title must be 1-120 characters" });
    return;
  }
  const todo = { id: nextId, title: req.body.title, completed: false };
  todos.set(nextId, todo);
  nextId += 1;
  res.status(201).json(todo);
});

router.patch("/todos/:todoId", (req, res) => {
  const todoId = Number(req.params.todoId);
  if (!todos.has(todoId)) {
    res.status(404).json({ detail: "Todo not found" });
    return;
  }
  if (typeof req.body.title !== "string" || req.body.title.trim() === "" || req.body.title.length > 120) {
    res.status(400).json({ detail: "Todo title must be 1-120 characters" });
    return;
  }
  const updated = { ...todos.get(todoId), title: req.body.title };
  todos.set(todoId, updated);
  res.json(updated);
});

module.exports = router;

File: sandbox/test_project/backend/app/api/routes/todos.py
from fastapi import APIRouter, HTTPException, status

from app.schemas.todo import TodoCreate, TodoRead

router = APIRouter(prefix="/todos", tags=["todos"])

_todos: dict[int, TodoRead] = {}
_next_id = 1


@router.get("", response_model=list[TodoRead])
def list_todos() -> list[TodoRead]:
    return list(_todos.values())


@router.post("", response_model=TodoRead, status_code=status.HTTP_201_CREATED)
def create_todo(payload: TodoCreate) -> TodoRead:
    global _next_id
    todo = TodoRead(id=_next_id, title=payload.title, completed=False)
    _todos[todo.id] = todo
    _next_id += 1
    return todo


@router.patch("/{todo_id}", response_model=TodoRead)
def update_todo(todo_id: int, payload: TodoCreate) -> TodoRead:
    if todo_id not in _todos:
        raise HTTPException(status_code=404, detail="Todo not found")
    existing = _todos[todo_id]
    updated = existing.model_copy(update={"title": payload.title})
    _todos[todo_id] = updated
    return updated

